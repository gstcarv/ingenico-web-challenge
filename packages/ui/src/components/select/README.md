# ðŸ“‹ Select Component

A customizable select dropdown component with support for prefixes, suffixes, and accessible design.

## ðŸ“¦ Installation

```bash
import { Select } from '@ingenico-challenge/ui'
```

## ðŸš€ Usage

### Basic Select

```tsx
import { Select } from '@ingenico-challenge/ui'

function MyComponent() {
  return (
    <Select>
      <Select.Option value="">Select an option</Select.Option>
      <Select.Option value="option1">Option 1</Select.Option>
      <Select.Option value="option2">Option 2</Select.Option>
      <Select.Option value="option3">Option 3</Select.Option>
    </Select>
  )
}
```

### Select with Prefix/Suffix

```tsx
// With prefix icon
<Select prefix={<CurrencyIcon />}>
  <Select.Option value="USD">USD</Select.Option>
  <Select.Option value="EUR">EUR</Select.Option>
  <Select.Option value="GBP">GBP</Select.Option>
</Select>

// With suffix (custom suffix will override default chevron)
<Select suffix={<CustomIcon />}>
  <Select.Option value="option1">Option 1</Select.Option>
</Select>
```

### Controlled Select

```tsx
import { useState } from 'react'
import { Select } from '@ingenico-challenge/ui'

function ControlledSelect() {
  const [value, setValue] = useState('')

  return (
    <Select value={value} onChange={(e) => setValue(e.target.value)}>
      <Select.Option value="">Select currency</Select.Option>
      <Select.Option value="USD">US Dollar</Select.Option>
      <Select.Option value="EUR">Euro</Select.Option>
      <Select.Option value="GBP">British Pound</Select.Option>
    </Select>
  )
}
```

### Select with Form Integration

```tsx
import { useForm } from 'react-hook-form'
import { Select } from '@ingenico-challenge/ui'

function FormSelect() {
  const { register } = useForm()

  return (
    <Select {...register('currency')}>
      <Select.Option value="">Select currency</Select.Option>
      <Select.Option value="USD">USD</Select.Option>
      <Select.Option value="EUR">EUR</Select.Option>
    </Select>
  )
}
```

## ðŸ“‹ Props

### Select Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `prefix` | `ReactNode` | - | Content to display before the select input |
| `suffix` | `ReactNode` | `<ChevronDown />` | Content to display after the select input |
| `className` | `string` | - | Additional CSS classes |
| `...props` | `SelectHTMLAttributes` | - | All standard select attributes |

### Select.Option Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `value` | `string` | - | Option value |
| `disabled` | `boolean` | `false` | Disables the option |
| `...props` | `OptionHTMLAttributes` | - | All standard option attributes |

## ðŸŽ¨ Styling

The component extends `InputBase` for consistent styling and supports:

- Custom prefix/suffix positioning
- Focus states with brand color ring
- Disabled state styling
- Responsive design
- Custom className overrides

### Default Styling
- Rounded corners (`rounded-lg`)
- Focus ring with brand primary color
- Chevron down icon on the right
- Proper spacing for prefix content

## â™¿ Accessibility

- Semantic HTML select element
- Proper focus management
- Keyboard navigation support
- Screen reader friendly
- ARIA attributes support

## ðŸ§ª Testing

```tsx
import { render, screen } from '@testing-library/react'
import { Select } from './select'

test('renders select with options', () => {
  render(
    <Select>
      <Select.Option value="option1">Option 1</Select.Option>
      <Select.Option value="option2">Option 2</Select.Option>
    </Select>
  )
  
  expect(screen.getByRole('combobox')).toBeInTheDocument()
  expect(screen.getByText('Option 1')).toBeInTheDocument()
  expect(screen.getByText('Option 2')).toBeInTheDocument()
})

test('handles value changes', () => {
  const handleChange = jest.fn()
  render(
    <Select onChange={handleChange}>
      <Select.Option value="option1">Option 1</Select.Option>
    </Select>
  )
  
  const select = screen.getByRole('combobox')
  select.value = 'option1'
  select.dispatchEvent(new Event('change'))
  
  expect(handleChange).toHaveBeenCalled()
})
```

## ðŸ“š Storybook

View interactive examples and documentation in Storybook:

```bash
cd packages/ui
pnpm storybook
```

Navigate to the Select component in the Storybook sidebar.

## ðŸ”§ Customization

### Custom Icons

```tsx
import { DollarSign, Globe } from 'lucide-react'

// Currency selector with dollar icon
<Select prefix={<DollarSign size={20} />}>
  <Select.Option value="USD">USD</Select.Option>
  <Select.Option value="EUR">EUR</Select.Option>
</Select>

// Language selector with globe icon
<Select prefix={<Globe size={20} />}>
  <Select.Option value="en">English</Select.Option>
  <Select.Option value="es">Spanish</Select.Option>
</Select>
```

### Styling Overrides

```tsx
<Select className="w-64 bg-gray-50">
  <Select.Option value="option1">Option 1</Select.Option>
</Select>
```
