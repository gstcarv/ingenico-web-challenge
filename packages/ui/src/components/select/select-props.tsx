import { ComponentProps, ReactNode } from "react";

export type SelectProps = Omit<ComponentProps<"select">, "prefix"> & {
    prefix?: ReactNode;
};

export type SelectOptionProps = ComponentProps<"option">;
