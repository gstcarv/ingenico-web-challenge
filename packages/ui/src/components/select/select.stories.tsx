import type { Meta, StoryObj } from "@storybook/react-vite";
import { Select } from "./select";
import { DollarSign } from "lucide-react";

const meta: Meta<typeof Select> = {
    title: "Components/Select",
    component: Select,
    parameters: {
        layout: "centered",
    },
    tags: ["autodocs"],
    argTypes: {
        disabled: {
            control: "boolean",
        },
    },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        children: [
            <Select.Option key="1" value="option1">
                Option 1
            </Select.Option>,
            <Select.Option key="2" value="option2">
                Option 2
            </Select.Option>,
            <Select.Option key="3" value="option3">
                Option 3
            </Select.Option>,
        ],
    },
};

export const Empty: Story = {
    args: {
        children: [],
    },
};

export const WithPrefix: Story = {
    render: () => (
        <Select prefix={<DollarSign size={16} />}>
            <Select.Option value="usd">USD - US Dollar</Select.Option>
            <Select.Option value="eur">EUR - Euro</Select.Option>
            <Select.Option value="gbp">GBP - British Pound</Select.Option>
        </Select>
    ),
};

export const WithTextPrefix: Story = {
    args: {
        prefix: "R$",
        children: [
            <Select.Option key="1" value="option1">
                Option 1
            </Select.Option>,
            <Select.Option key="2" value="option2">
                Option 2
            </Select.Option>,
            <Select.Option key="3" value="option3">
                Option 3
            </Select.Option>,
        ],
    },
};

export const WithBrazilFlag: Story = {
    args: {
        prefix: "ðŸ‡§ðŸ‡·",
        children: [
            <Select.Option key="1" value="br">
                Brasil
            </Select.Option>,
            <Select.Option key="2" value="us">
                Estados Unidos
            </Select.Option>,
            <Select.Option key="3" value="pt">
                Portugal
            </Select.Option>,
        ],
    },
};
