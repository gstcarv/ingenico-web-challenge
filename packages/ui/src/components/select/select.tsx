import { cva } from "class-variance-authority";
import { ChevronDown } from "lucide-react";
import { SelectOptionProps, SelectProps } from "./select-props";
import { cn } from "../../utils";

const inputContainerVariants = cva(
    "h-16 w-70 bg-white border border-neutral-200 rounded-lg shadow-lg relative text-neutral-500 font-medium",
    {
        variants: {
            variant: {
                default: "",
            },
        },
    }
);

export const Select = ({ className, prefix, ...props }: SelectProps) => (
    <div className={cn(inputContainerVariants({ variant: "default" }), className)}>
        {prefix && (
            <div className="absolute left-4 top-1/2 transform -translate-y-1/2 pointer-events-none text-neutral-500">
                {prefix}
            </div>
        )}

        <select
            {...props}
            className={cn("select__input w-full h-full appearance-none bg-transparent", prefix ? "pl-12" : "px-4")}
        />

        <ChevronDown
            className="select__input-icon absolute right-3 top-1/2 transform -translate-y-1/2 pointer-events-none text-neutral-500"
            size={20}
        />
    </div>
);

Select.Option = (props: SelectOptionProps) => <option {...props} />;
