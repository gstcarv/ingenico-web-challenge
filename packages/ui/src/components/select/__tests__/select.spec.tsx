import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import { Select } from "../select";

describe("Select", () => {
    it("renders select element with correct attributes", () => {
        render(
            <Select data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
                <Select.Option value="option2">Option 2</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toBeInTheDocument();
        expect(selectElement.tagName).toBe("SELECT");
    });

    it("renders options correctly", () => {
        render(
            <Select>
                <Select.Option value="option1">Option 1</Select.Option>
                <Select.Option value="option2">Option 2</Select.Option>
                <Select.Option value="option3">Option 3</Select.Option>
            </Select>
        );

        expect(screen.getByText("Option 1")).toBeInTheDocument();
        expect(screen.getByText("Option 2")).toBeInTheDocument();
        expect(screen.getByText("Option 3")).toBeInTheDocument();
    });

    it("applies custom className", () => {
        const { container } = render(
            <Select className="custom-class" data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("renders prefix when provided", () => {
        render(
            <Select prefix="R$" data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        expect(screen.getByText("R$")).toBeInTheDocument();
    });

    it("renders React node prefix", () => {
        const TestIcon = () => <span data-testid="icon">ðŸ’°</span>;

        render(
            <Select prefix={<TestIcon />} data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        expect(screen.getByTestId("icon")).toBeInTheDocument();
    });

    it("applies correct padding when prefix is present", () => {
        const { container } = render(
            <Select prefix="R$" data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("applies default padding when no prefix is present", () => {
        const { container } = render(
            <Select data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("handles change events", () => {
        const handleChange = vi.fn();

        render(
            <Select onChange={handleChange} data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
                <Select.Option value="option2">Option 2</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        fireEvent.change(selectElement, { target: { value: "option2" } });

        expect(handleChange).toHaveBeenCalledTimes(1);
    });

    it("handles disabled state", () => {
        render(
            <Select disabled data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toBeDisabled();
    });

    it("handles value prop", () => {
        const handleChange = vi.fn();

        render(
            <Select value="option2" onChange={handleChange} data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
                <Select.Option value="option2">Option 2</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toHaveValue("option2");
    });

    it("handles multiple prop", () => {
        render(
            <Select multiple data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
                <Select.Option value="option2">Option 2</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toHaveAttribute("multiple");
    });

    it("renders chevron down icon", () => {
        render(
            <Select data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        const chevronIcon = screen.getByTestId("select").parentElement?.querySelector("svg");
        expect(chevronIcon).toBeInTheDocument();
    });

    it("applies correct base classes to container", () => {
        const { container } = render(
            <Select data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("applies correct classes to select element", () => {
        const { container } = render(
            <Select data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
            </Select>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("handles empty children", () => {
        render(<Select data-testid="select" />);

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toBeInTheDocument();
        expect(selectElement.children).toHaveLength(0);
    });

    it("forwards all HTML select attributes", () => {
        render(
            <Select data-testid="select" name="currency" id="currency-select" required aria-label="Select currency">
                <Select.Option value="usd">USD</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toHaveAttribute("name", "currency");
        expect(selectElement).toHaveAttribute("id", "currency-select");
        expect(selectElement).toHaveAttribute("required");
        expect(selectElement).toHaveAttribute("aria-label", "Select currency");
    });
});

describe("Select.Option", () => {
    it("renders option element with correct attributes", () => {
        render(
            <Select data-testid="select">
                <Select.Option value="option1" data-testid="option">
                    Option 1
                </Select.Option>
            </Select>
        );

        const optionElement = screen.getByTestId("option");
        expect(optionElement).toBeInTheDocument();
        expect(optionElement.tagName).toBe("OPTION");
        expect(optionElement).toHaveAttribute("value", "option1");
    });

    it("handles disabled option", () => {
        render(
            <Select data-testid="select">
                <Select.Option value="option1" disabled data-testid="option">
                    Option 1
                </Select.Option>
            </Select>
        );

        const optionElement = screen.getByTestId("option");
        expect(optionElement).toBeDisabled();
    });

    it("forwards all HTML option attributes", () => {
        render(
            <Select data-testid="select">
                <Select.Option value="option1" data-testid="option" aria-label="First option">
                    Option 1
                </Select.Option>
            </Select>
        );

        const optionElement = screen.getByTestId("option");
        expect(optionElement).toHaveAttribute("aria-label", "First option");
    });

    it("handles option with empty value", () => {
        render(
            <Select data-testid="select">
                <Select.Option value="" data-testid="option">
                    Empty Option
                </Select.Option>
            </Select>
        );

        const optionElement = screen.getByTestId("option");
        expect(optionElement).toHaveAttribute("value", "");
        expect(optionElement).toHaveTextContent("Empty Option");
    });
});

describe("Select Accessibility", () => {
    it("supports aria-label", () => {
        render(
            <Select aria-label="Select currency" data-testid="select">
                <Select.Option value="usd">USD</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toHaveAttribute("aria-label", "Select currency");
    });

    it("supports aria-describedby", () => {
        render(
            <Select aria-describedby="description" data-testid="select">
                <Select.Option value="usd">USD</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toHaveAttribute("aria-describedby", "description");
    });

    it("supports role attribute", () => {
        render(
            <Select role="combobox" data-testid="select">
                <Select.Option value="usd">USD</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");
        expect(selectElement).toHaveAttribute("role", "combobox");
    });
});

describe("Select Integration", () => {
    it("works with form elements", () => {
        const handleSubmit = vi.fn((e) => e.preventDefault());

        render(
            <form onSubmit={handleSubmit}>
                <Select name="currency" data-testid="select">
                    <Select.Option value="usd">USD</Select.Option>
                    <Select.Option value="eur">EUR</Select.Option>
                </Select>
                <button type="submit">Submit</button>
            </form>
        );

        const selectElement = screen.getByTestId("select");
        const submitButton = screen.getByText("Submit");

        fireEvent.change(selectElement, { target: { value: "eur" } });
        fireEvent.click(submitButton);

        expect(handleSubmit).toHaveBeenCalledTimes(1);
    });

    it("handles keyboard navigation", () => {
        const handleChange = vi.fn();

        render(
            <Select onChange={handleChange} data-testid="select">
                <Select.Option value="option1">Option 1</Select.Option>
                <Select.Option value="option2">Option 2</Select.Option>
            </Select>
        );

        const selectElement = screen.getByTestId("select");

        // Focus the select
        selectElement.focus();
        expect(selectElement).toHaveFocus();

        // Simulate keyboard interaction
        fireEvent.keyDown(selectElement, { key: "ArrowDown" });
        fireEvent.change(selectElement, { target: { value: "option2" } });

        expect(handleChange).toHaveBeenCalledTimes(1);
    });
});
