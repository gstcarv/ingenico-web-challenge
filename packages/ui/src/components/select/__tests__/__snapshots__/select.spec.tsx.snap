// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > applies correct base classes to container 1`] = `
<div
  class="h-16 w-70 bg-white border border-neutral-200 rounded-lg shadow-lg relative text-neutral-500 font-medium"
>
  <select
    class="select__input w-full h-full appearance-none bg-transparent px-4"
    data-testid="select"
  >
    <option
      value="option1"
    >
      Option 1
    </option>
  </select>
  <svg
    aria-hidden="true"
    class="lucide lucide-chevron-down select__input-icon absolute right-3 top-1/2 transform -translate-y-1/2 pointer-events-none text-neutral-500"
    fill="none"
    height="20"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m6 9 6 6 6-6"
    />
  </svg>
</div>
`;

exports[`Select > applies correct classes to select element 1`] = `
<div
  class="h-16 w-70 bg-white border border-neutral-200 rounded-lg shadow-lg relative text-neutral-500 font-medium"
>
  <select
    class="select__input w-full h-full appearance-none bg-transparent px-4"
    data-testid="select"
  >
    <option
      value="option1"
    >
      Option 1
    </option>
  </select>
  <svg
    aria-hidden="true"
    class="lucide lucide-chevron-down select__input-icon absolute right-3 top-1/2 transform -translate-y-1/2 pointer-events-none text-neutral-500"
    fill="none"
    height="20"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m6 9 6 6 6-6"
    />
  </svg>
</div>
`;

exports[`Select > applies correct padding when prefix is present 1`] = `
<div
  class="h-16 w-70 bg-white border border-neutral-200 rounded-lg shadow-lg relative text-neutral-500 font-medium"
>
  <div
    class="absolute left-4 top-1/2 transform -translate-y-1/2 pointer-events-none text-neutral-500"
  >
    R$
  </div>
  <select
    class="select__input w-full h-full appearance-none bg-transparent pl-12"
    data-testid="select"
  >
    <option
      value="option1"
    >
      Option 1
    </option>
  </select>
  <svg
    aria-hidden="true"
    class="lucide lucide-chevron-down select__input-icon absolute right-3 top-1/2 transform -translate-y-1/2 pointer-events-none text-neutral-500"
    fill="none"
    height="20"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m6 9 6 6 6-6"
    />
  </svg>
</div>
`;

exports[`Select > applies custom className 1`] = `
<div
  class="h-16 w-70 bg-white border border-neutral-200 rounded-lg shadow-lg relative text-neutral-500 font-medium custom-class"
>
  <select
    class="select__input w-full h-full appearance-none bg-transparent px-4"
    data-testid="select"
  >
    <option
      value="option1"
    >
      Option 1
    </option>
  </select>
  <svg
    aria-hidden="true"
    class="lucide lucide-chevron-down select__input-icon absolute right-3 top-1/2 transform -translate-y-1/2 pointer-events-none text-neutral-500"
    fill="none"
    height="20"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m6 9 6 6 6-6"
    />
  </svg>
</div>
`;

exports[`Select > applies default padding when no prefix is present 1`] = `
<div
  class="h-16 w-70 bg-white border border-neutral-200 rounded-lg shadow-lg relative text-neutral-500 font-medium"
>
  <select
    class="select__input w-full h-full appearance-none bg-transparent px-4"
    data-testid="select"
  >
    <option
      value="option1"
    >
      Option 1
    </option>
  </select>
  <svg
    aria-hidden="true"
    class="lucide lucide-chevron-down select__input-icon absolute right-3 top-1/2 transform -translate-y-1/2 pointer-events-none text-neutral-500"
    fill="none"
    height="20"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m6 9 6 6 6-6"
    />
  </svg>
</div>
`;
